lark/parsers/earley.py

```diff
@@ -9,34 +9,36 @@
 is explained here: https://lark-parser.readthedocs.io/en/latest/_static/sppf/sppf.html
 """
 
-import typing
-
+from typing import TYPE_CHECKING, Callable
 from collections import deque
 
 from ..lexer import Token
 from ..tree import Tree
 from ..exceptions import UnexpectedEOF, UnexpectedToken
-from ..utils import logger
+from ..utils import logger, OrderedSet
 from .grammar_analysis import GrammarAnalyzer
 from ..grammar import NonTerminal
 from .earley_common import Item
-from .earley_forest import ForestSumVisitor, SymbolNode, TokenNode, ForestToParseTree
+from .earley_forest import ForestSumVisitor, SymbolNode, StableSymbolNode, TokenNode, ForestToParseTree
 
-if typing.TYPE_CHECKING:
+if TYPE_CHECKING:
     from ..common import LexerConf, ParserConf
 
 class Parser:
     lexer_conf: 'LexerConf'
     parser_conf: 'ParserConf'
     debug: bool
 
-    def __init__(self, lexer_conf: 'LexerConf', parser_conf: 'ParserConf', term_matcher, resolve_ambiguity=True, debug=False, tree_class=Tree):
+    def __init__(self, lexer_conf: 'LexerConf', parser_conf: 'ParserConf', term_matcher: Callable,
+                 resolve_ambiguity: bool=True, debug: bool=False, tree_class: type=Tree, ordered_sets: bool=True):
```

Comment:
Note: In `LarkOptions` you defined `tree_class` as `Callable`. 
A more specific type might be `Type[Tree]` if we require that tree classes be subclasses of `Tree`. Also, `tree_class` here is `Optional[]`.

Reply:
Suggestion: Don't require it to be a subclass of `Tree`.

And I don't think `tree_class` is `Optional`? It can't be `None`, that would cause an error, no?

Reply:
Isn't `tree_class` defined as Any in `LarkOptions`?

It could be Callable, I agree with that. Demanding that it's an instance of Tree might be a bit too limiting. Ideally, imho, we would use Protocol, once we drop Python 3.7.

Anyway, I think Callable is good enough for now, since this PR doesn't pretend to fix *all* the types.

Reply:
Changed

Reply:
> And I don't think `tree_class` is `Optional`? It can't be `None`, that would cause an error, no?

`tree_class` is `None` for `ambiguity == 'forest'`:

https://github.com/lark-parser/lark/blob/c2e44ea7f823b76695b58304056a4178fadbc873/lark/parser_frontends.py#L200

https://github.com/lark-parser/lark/blob/c2e44ea7f823b76695b58304056a4178fadbc873/lark/parsers/earley.py#L299-L305

mypy may not be reporting an error because `tree_class` gets passed to the `Parser` constructor via dictionary unpacking. i.e.

https://github.com/lark-parser/lark/blob/c2e44ea7f823b76695b58304056a4178fadbc873/lark/parser_frontends.py#L194-L195

> We can turn it into `Callable[[str, Sequence], Any]` for now. Does that make sense?

That works

Reply:
When you're right, you're right!